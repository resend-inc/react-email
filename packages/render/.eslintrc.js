module.exports = {
  extends: ["custom/react-internal"],
  rules: {
    "import/no-default-export": "off",
    "@typescript-eslint/consistent-type-imports": "off",
    "tsdoc/syntax": "off"
  },
};
