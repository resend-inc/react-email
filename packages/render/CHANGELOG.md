# @react-email/render

## 0.0.17
