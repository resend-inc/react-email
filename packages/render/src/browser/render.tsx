import { convert } from "html-to-text";
import { Suspense } from "react";
import { renderToReadableStream } from "react-dom/server";
import { Options } from "../shared/options";
import { plainTextSelectors } from "../shared/plain-text-selectors";
import { pretty } from "../shared/pretty";

async function streamToString(
  stream: ReadableStream<Uint8Array>,
): Promise<string> {
  const reader = stream.getReader();
  const decoder = new TextDecoder("utf-8");
  let result = "";
  let done: boolean | undefined;

  while (!done) {
    // eslint-disable-next-line no-await-in-loop
    const { value, done: doneReading } = await reader.read();
    done = doneReading;
    if (value) {
      result += decoder.decode(value, { stream: !done });
    }
  }

  return result;
}

export async function render(
  element: React.ReactElement,
  options?: Options,
): Promise<string> {
  const suspendedElement = <Suspense>{element}</Suspense>;
<<<<<<< HEAD
  const reactDOMServer = await import("react-dom/server");

  let html!: string;
  if (Object.hasOwn(reactDOMServer, "renderToReadableStream")) {
    html = await readStream(
      await reactDOMServer.renderToReadableStream(suspendedElement),
    );
  } else {
    await new Promise<void>((resolve, reject) => {
      const stream = reactDOMServer.renderToPipeableStream(suspendedElement, {
        async onAllReady() {
          html = await readStream(stream);
          resolve();
        },
        onError(error) {
          reject(error as Error);
        },
      });
    });
  }
=======
  const readableStream = await renderToReadableStream(suspendedElement);
  const html = await streamToString(readableStream);
>>>>>>> 447c7780 (refactor: `render` package)

  if (options?.plainText) {
    return convert(html, {
      selectors: plainTextSelectors,
      ...options.htmlToTextOptions,
    });
  }

  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';

  const document = `${doctype}${html.replace(/<!DOCTYPE.*?>/, "")}`;

  if (options?.pretty) {
    return pretty(document);
  }

  return document;
}
