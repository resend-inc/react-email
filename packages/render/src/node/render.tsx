import { Writable } from "node:stream";
import { renderToPipeableStream } from "react-dom/server";
import { Suspense } from "react";
import { convert } from "html-to-text";
import { Options } from "../shared/options";
import { pretty } from "../shared/pretty";
import { plainTextSelectors } from "../shared/plain-text-selectors";

export async function render(
  element: React.ReactElement,
  options?: Options,
): Promise<string> {
  const suspendedElement = <Suspense>{element}</Suspense>;
  const html = await new Promise<string>((resolve, reject) => {
    let htmlChunksString = "";

    const writable = new Writable({
      write(chunk: Buffer, _encoding, callback) {
        htmlChunksString += chunk.toString();
        callback();
      },
    });

    writable.on("finish", () => {
      resolve(htmlChunksString);
    });

    writable.on("error", (err) => {
      reject(err);
    });

    const stream = renderToPipeableStream(suspendedElement, {
      onAllReady() {
        stream.pipe(writable);
      },
      onError(error) {
        reject(error as Error);
      },
    });
  });

  if (options?.plainText) {
    return convert(html, {
      selectors: plainTextSelectors,
      ...options.htmlToTextOptions,
    });
  }

  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';

  const document = `${doctype}${html.replace(/<!DOCTYPE.*?>/, "")}`;

  if (options?.pretty) {
    return pretty(document);
  }

  return document;
}
