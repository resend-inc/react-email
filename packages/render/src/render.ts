import * as ReactDomServer from "react-dom/server";
import { convert } from "html-to-text";
import pretty from "pretty";

export interface Options {
  pretty?: boolean;
  plainText?: boolean;
}

export const render = (component: React.ReactElement, options?: Options) => {
  const markup = ReactDomServer.renderToStaticMarkup(component);
  if (options?.plainText) {
    return convertToPlainText(markup);
  }
  
  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';
  const document = `${doctype}${markup}`;

  if (options && options.pretty) {
    return pretty(document);
  }

  return document;
};

export const convertToPlainText = (staticMarkup: string) => {
  return convert(staticMarkup, {
    selectors: [
      { selector: "img", format: "skip" },
      { selector: "#__react-email-preview", format: "skip" },
    ],
  });
};