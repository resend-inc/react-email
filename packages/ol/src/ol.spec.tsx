import { Ol } from './index';
import { render } from '@react-email/render';

describe('render', () => {
  beforeEach(() => {
    jest.restoreAllMocks();
    jest.resetModules();
  });

  it('renders the <Ol> component', () => {
    const actualOutput = render(<Ol />);
    expect(actualOutput).toMatchSnapshot();
  });

  it('renders children components', () => {
    const actualOutput = render(
      <Ol>
        <li>Item 1</li>
      </Ol>,
    );
    expect(actualOutput).toMatchSnapshot();
  });
});
