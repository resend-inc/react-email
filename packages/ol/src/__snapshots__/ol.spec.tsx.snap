// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders children components 1`] = `"<!DOCTYPE html PUBLIC \\"-//W3C//DTD XHTML 1.0 Transitional//EN\\" \\"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\\"><ol style=\\"list-style-type:upper-roman\\"><li>Item 1</li></ol>"`;

exports[`render renders the <Ol> component 1`] = `"<!DOCTYPE html PUBLIC \\"-//W3C//DTD XHTML 1.0 Transitional//EN\\" \\"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\\"><ol style=\\"list-style-type:upper-roman\\"></ol>"`;
